src/OpenTelemetry/Metrics/Exemplar/FixedSizeExemplarReservoir.cs

```diff
@@ -0,0 +1,146 @@
+
+        Debug.Assert(activeBuffer != null, "activeBuffer was null");
+
+        var inactiveBuffer = activeBuffer == this.bufferA
+            ? this.bufferB
+            : this.bufferA;
+
+        if (this.ResetOnCollect)
+        {
+            for (int i = 0; i < inactiveBuffer.Length; i++)
+            {
+                inactiveBuffer[i].Reset();
+            }
+
+            this.OnReset();
+        }
+        else
+        {
+#if NET6_0_OR_GREATER
+            var length = this.Capacity;
+            ref var inactive = ref MemoryMarshal.GetArrayDataReference(inactiveBuffer);
+            ref var active = ref MemoryMarshal.GetArrayDataReference(activeBuffer);
+            do
+            {
+                if (active.IsUpdated())
+                {
+                    active.Copy(ref inactive);
+                }
+
+                inactive = ref Unsafe.Add(ref inactive, 1);
+                active = ref Unsafe.Add(ref active, 1);
+            }
+            while (--length > 0);
+#else
+            for (int i = 0; i < activeBuffer!.Length; i++)
+            {
+                ref var active = ref activeBuffer[i];
+                if (active.IsUpdated())
+                {
+                    active.Copy(ref inactiveBuffer[i]);
+                }
+            }
+#endif
+        }
+
+        Interlocked.Exchange(ref this.activeBuffer, inactiveBuffer);
+
+        return new(activeBuffer!);
+    }
+
+    internal sealed override void Initialize(AggregatorStore aggregatorStore)
+    {
+        var viewDefinedTagKeys = aggregatorStore.TagKeysInteresting;
+
+#if NET6_0_OR_GREATER
+        var length = this.bufferA.Length;
+        ref var a = ref MemoryMarshal.GetArrayDataReference(this.bufferA);
+        ref var b = ref MemoryMarshal.GetArrayDataReference(this.bufferB);
+        do
+        {
+            a.ViewDefinedTagKeys = viewDefinedTagKeys;
+            b.ViewDefinedTagKeys = viewDefinedTagKeys;
+            a = ref Unsafe.Add(ref a, 1);
+            b = ref Unsafe.Add(ref b, 1);
+        }
+        while (--length > 0);
+#else
+        for (int i = 0;
+            i < this.bufferA.Length && i < this.bufferB.Length;
+            i++)
+        {
+            this.bufferA[i].ViewDefinedTagKeys = viewDefinedTagKeys;
+            this.bufferB[i].ViewDefinedTagKeys = viewDefinedTagKeys;
+        }
+#endif
+
+        base.Initialize(aggregatorStore);
+    }
+
+    protected virtual void OnReset()
+    {
+    }
+
+    protected void UpdateExemplar<T>(int exemplarIndex, in ExemplarMeasurement<T> measurement)
+        where T : struct
+    {
+        var activeBuffer = Interlocked.CompareExchange(ref this.activeBuffer, null, NeverMatchExemplarBuffer)
+            ?? this.AcquireActiveBufferRare();
+
+        activeBuffer[exemplarIndex].Update(in measurement);
+    }
+
+    private Exemplar[] AcquireActiveBufferRare()
+    {
+        // Note: We reach here if performing a write while racing with collect.
```

Comment:
one thing I noticed with this active-passive buffer idea is that - an exemplar update, when competing with collect(), gets pushed to the next cycle.
i.e, say a measurement that occurred at T9, if stored in exemplar will be exported with T1-T10 timeinterval previously. With this PR, the exemplar at T9 could be exported with the T10-T20 batch.. Given Exemplar stores the actual timestamp, backends should be able to deal with this, so I am not seeing this as an issue. (Also not aware of any OSS delta backends to test this functionality either.)

Reply:
There is a little bit of nuance to the current implementation just FYI. If an exemplar is able to obtain the active buffer AND [obtain the update lock](https://github.com/open-telemetry/opentelemetry-dotnet/blob/4512b49465cb53d5a4f97841286ad2c1f7b090e8/src/OpenTelemetry/Metrics/Exemplar/Exemplar.cs#L99-L103) then during export we will actually [wait for that update to complete](https://github.com/open-telemetry/opentelemetry-dotnet/blob/4512b49465cb53d5a4f97841286ad2c1f7b090e8/src/OpenTelemetry/Metrics/Exemplar/ReadOnlyExemplarCollection.cs#L111) so it will become part of the same collection cycle!

Now that "AND" part is a problem. There is essentially a race there. I have some idea how to fix it but hoping to do that on a follow-up 😄

Reply:
Yes. Lets put a TODO, and not block this PR.